tests/layer/test_cluster_gcn.py

```diff
@@ -0,0 +1,241 @@
+    assert conf["kernel_constraint"] == None
+    assert conf["bias_constraint"] == None
+
+
+def test_ClusterGraphConvolution_init():
+    cluster_gcn_layer = ClusterGraphConvolution(units=16, activation="relu")
+
+    assert cluster_gcn_layer.units == 16
+    assert cluster_gcn_layer.use_bias == True
+    assert cluster_gcn_layer.get_config()["activation"] == "relu"
+
+
+def test_GraphConvolution():
+    G, features = create_graph_features()
+
+    # We need to specify the batch shape as one for the ClusterGraphConvolutional logic to work
+    x_t = Input(batch_shape=(1,) + features.shape, name="X")
+    A_t = Input(batch_shape=(1, 3, 3), name="A")
+    output_indices_t = Input(batch_shape=(1, None), dtype="int32", name="outind")
+
+    # Note we add a batch dimension of 1 to model inputs
+    adj = nx.to_numpy_array(G)[None, :, :]
+    out_indices = np.array([[0, 1]], dtype="int32")
+    x = features[None, :, :]
+
+    # Remove batch dimension
+    A_mat = Lambda(lambda A: K.squeeze(A, 0))(A_t)
+
+    # Test with final_layer=False
+    out = ClusterGraphConvolution(2, final_layer=False)([x_t, output_indices_t, A_mat])
+    model = keras.Model(inputs=[x_t, A_t, output_indices_t], outputs=out)
+    preds = model.predict([x, adj, out_indices], batch_size=1)
+    assert preds.shape == (1, 3, 2)
+
+    # Now try with final_layer=True
+    out = ClusterGraphConvolution(2, final_layer=True)([x_t, output_indices_t, A_mat])
+    # The final layer removes the batch dimension and causes the call to predict to fail.
+    # We are going to manually added the batch dimension before calling predict.
+    out = K.expand_dims(out, 0)
+    model = keras.Model(inputs=[x_t, A_t, output_indices_t], outputs=out)
+    print(
+        f"x_t: {x_t.shape} A_t: {A_t.shape} output_indices_t: {output_indices_t.shape}"
+    )
+    preds = model.predict([x, adj, out_indices], batch_size=1)
+    assert preds.shape == (1, 2, 2)
+
+    # Check for errors with batch size != 1
+    # We need to specify the batch shape as one for the ClusterGraphConvolutional logic to work
+    x_t = Input(batch_shape=(2,) + features.shape)
+    output_indices_t = Input(batch_shape=(2, None), dtype="int32")
+    with pytest.raises(ValueError):
+        out = ClusterGraphConvolution(2)([x_t, A_t, output_indices_t])
+
+
+def test_ClusterGCN_init():
+    G, features = create_graph_features()
+    nodes = G.nodes()
+    node_features = pd.DataFrame.from_dict(
+        {n: f for n, f in zip(nodes, features)}, orient="index"
+    )
+    G = StellarGraph(G, node_type_name="node", node_features=node_features)
+
+    generator = ClusterNodeGenerator(G)
+    cluster_gcn_model = ClusterGCN(
+        layer_sizes=[2], generator=generator, activations=["relu"], dropout=0.5
+    )
+
+    assert cluster_gcn_model.layer_sizes == [2]
+    assert cluster_gcn_model.activations == ["relu"]
+    assert cluster_gcn_model.dropout == 0.5
+
+
+def test_ClusterGCN_apply():
+    G, features = create_graph_features()
```

Comment:
Given that this block of code is (mostly) reused multiple times, I would encapsulate it along with a switch for computing `adj`, e.g.: ```def create_stellargraph(with_adj=False):
    Gnx, features = create_graph_features()
    nodes = Gnx.nodes()
    node_features = pd.DataFrame.from_dict(
        {n: f for n, f in zip(nodes, features)}, orient="index"
    )
    G = StellarGraph(Gnx, node_features=node_features)
    if with_adj:
        adj = nx.to_numpy_array(Gnx)[None, :, :]
        return G, adj
    return G```

Reply:
I've simplified it. Thank you!